Cluster wait

idorrington92 opened the pull request:
Closes #6346 

- [x] Tests added / passed
- [x] Passes `pre-commit run --all-files`

GPUtester commented:
Can one of the admins verify this patch?

quasiben commented:
add to allowlist

idorrington92 commented:
> add to allowlist

@quasiben Sorry, what does this mean?

Also, I've got 3 failing tests in test_ssh. Is it possible to run these locally? They just fail for me, saying cluster failed to start. This happens on main branch as well. I'm not really sure how to debug them otherwise...

quasiben commented:
@idorrington92 the comment allows us to run your PR against a GPU CI system:
https://docs.dask.org/en/stable/develop.html#gpu-ci

Distributed has been flaky in the past and recently there has been a concerted effort to stabilize the CI test:
https://dask.github.io/distributed/test_report.html

As you can see in the linked report, `test_dashboard_port_zero` is a known flaky test

I would suggest @jacobtomlinson review this PR but he is currently at SciPy (as are other dask maintainers) so it may be a little bit before you hear back

jacobtomlinson reviewed (commented):
Many thanks for raising this!

idorrington92 commented on distributed/client.py:
```diff
@@ -1332,7 +1332,10 @@ def wait_for_workers(self, n_workers=0, timeout=None):
         """
-        return self.sync(self._wait_for_workers, n_workers, timeout=timeout)
+        if self.cluster is not None:
+            return self.cluster.wait_for_workers(n_workers, timeout)
```
I think I've fixed it, if I understood you correctly

jacobtomlinson reviewed (commented):
Thanks for addressing the review feedback.

It looks like the linter isn't happy. [If you haven't installed `pre-commit` you'll want to do that](https://docs.dask.org/en/stable/develop.html#code-formatting).

Also it looks like some of the test failures might be related, especially this one.

```python-traceback
_____________________ test_ssh_nprocs_renamed_to_n_workers _____________________
/usr/share/miniconda3/envs/dask-distributed/lib/python3.8/site-packages/coverage/data.py:130: CoverageWarning: Data file '/home/runner/work/distributed/distributed/.coverage.fv-az302-754.15681.346903' doesn't seem to be a coverage data file: cannot unpack non-iterable NoneType object
  data._warn(str(exc))
@gen_test()
asyncdeftest_ssh_nprocs_renamed_to_n_workers():
with pytest.warns(FutureWarning, match="renamed to n_workers"):
asyncwith SSHCluster(
                ["127.0.0.1"] * 3,
                connect_options=dict(known_hosts=None),
                asynchronous=True,
                scheduler_options={"idle_timeout": "5s"},
                worker_options={"death_timeout": "5s", "nprocs": 2},
            ) as cluster:
assertlen(cluster.workers) == 2
asyncwith Client(cluster, asynchronous=True) as client:
>                   await client.wait_for_workers(4)
distributed/deploy/tests/test_ssh.py:117: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
self = SSHCluster(SSHCluster, 'tcp://10.1.0.103:42191', workers=0, threads=0, memory=0 B)
n_workers = 4, timeout = None
asyncdef_wait_for_workers(self, n_workers=0, timeout=None):
>       info = self.scheduler.identity()
E       AttributeError: 'Scheduler' object has no attribute 'identity'
distributed/deploy/cluster.py:534: AttributeError
```

Could you take a look?

idorrington92 commented:
Doesn't it say listing/pre-commit hooks passed ok? 

I agree that test failure is related but I can't run it locally as I get an error saying cluster failed to start. 

Is there something I need to do in order to run it locally? Otherwise the only thing I've got to go on is the automated tests here...

jacobtomlinson commented:
Apologies I was looking at `Linting / Unit Test Results (pull_request)` which actually has nothing to do with linting, sorry for the noise.

Can you run `ssh localhost` without having to enter a password?

idorrington92 commented:
Thanks @jacobtomlinson, that fixed the error I was getting. I would have taken ages to think of that.

I had a little look at the bug that was coming up in the tests, but it's not obvious to me why it's happening. I'll have proper look later in the week.

idorrington92 commented:
@jacobtomlinson can I get some help please? 

I managed to fix those failing tests by calling client.scale in client.wait_for_workers. But when I merged in master, the test I wrote to check the cluster.wait_for_worker method started failing. It waits forever for the workers because they never change their status to running, they just get stuck on init. 

Do you have any ideas why this would be happening? No worries if not, I'll keep digging.

fjetter commented on distributed/client.py:
```diff
@@ -1367,7 +1367,13 @@ def wait_for_workers(self, n_workers=0, timeout=None):
         """
-        return self.sync(self._wait_for_workers, n_workers, timeout=timeout)
+        try:
+            self.scale(n_workers)
```
I think adding this would quite dramatically change behavior of this API. Is this intentional? https://github.com/dask/distributed/issues/6346 doesn't mention this

idorrington92 replied:
You're right, this is a mistake. 

I added this to get test_n_workers in test_ssh.py to work. The test looks like it starts a cluster with 2 workers and then waits for 4 to be running, which never happened so I scaled up to 4 workers. I don't understand how the code was, before I added the call to scale, functionally any different to main branch. I also don't understand how that test passes on main if it is waiting for more workers than it has, wouldn't it be waiting forever?

idorrington92 replied:
I have removed this call to scale. The problem I was seeing was due to cluster.scheduler_info not updating. I thought it was always updated, but it turns out it only updates when certain commands are sent to the scheduler, including the call to add workers. Calling scale adds workers and updates the scheduler_info, so it fixed the bugs I was seeing. Now I manually update the scheduler_info, fixing the original issue and meaning this call to scale is no longer necessary (and as you said, was a dramatic change of behaviour anyway). 

Thanks for your comment, it forced me to look more deeply into the problem and find a better solution

idorrington92 commented:
I think it's all working properly now. I had to create a function to get the worker status and update the scheduler_info. As it was, scheduler_info was only updated when the workers were added, which was why it looked like they were stuck with status 'init'. The new function updates the scheduler_info with the latest worker status. 

One test is failing on python 3.8, but I think that's nothing to do with my changes.

jacobtomlinson commented on distributed/deploy/cluster.py:
```diff
@@ -564,3 +565,51 @@ def __eq__(self, other):
+            info = await self.get_worker_status()
+            self._scheduler_identity = info
+
+    async def get_worker_status(self):
+        comm = await self.scheduler_comm.live_comm()
+        comm.name = "Cluster worker status"
+        await comm.write({"op": "identity"})
+        self.scheduler_info = await comm.read()
+        return SchedulerInfo(self.scheduler_info)
```
This feels a little muddy as it is updating the `scheduler_info` which is then used to read the worker status outside of the function. There is also a bunch of duplication here from the `_start` method. Perhaps this should be renamed to `update_scheduler_info` or similar and used in both places.

Also given that it opens a comm and uses it briefly perhaps it would be better to use a context manager?

idorrington92 replied:
Yes, it's much better after a little refactoring.

I tried to use a context manager, but I don't think it works here as the cluster needs to keep scheduler_comm open or we start getting errors in other places, e.g. I was getting a RuntimeError during startup when it called get_metadata. Or did I misunderstand?

jacobtomlinson commented on distributed/deploy/cluster.py:
```diff
@@ -564,3 +565,71 @@ def __eq__(self, other):
+        comm.name = comm_name
+        await comm.write({"op": op})
+        self.scheduler_info = await comm.read()
+        return comm
```
This still feels strange, why would a method called `update_scheduler_info` take a `comm` and return a `live_comm` that then gets reused by the `_watch_worker_status`?

I the `_start` we should create the comm and do whatever we need to do with it. Then in `update_scheduler_info` we should just use that existing comm. Maybe part of the issue here is that we don't assign `comm` to anything in `_start` so can't reuse it later.

idorrington92 replied:
I tried assigning comm to an attribute and reusing it, but it raised an assertion error about reading for a comm that was already being read, I assume as it's being used to wait for workers being added/removed. Instead I created a new comm in _start and use that whenever we need to update scheduler_info. It seems to work fine, and update_scheduler_info is now much simpler. Let me know what you think

jacobtomlinson replied:
Thanks for tackling this, it definitely reads a bit more clearly now.

It'd be great to get a review from @fjetter here on the implications of handling multiple comms this way.

idorrington92 commented:
Hi @jacobtomlinson and @fjetter, just pinging you to make sure this isn't forgotten about :) 

I've merged main branch into this branch, and copied the new behaviour for when there are no workers

jacobtomlinson reviewed (approved):
I'm happy with this PR, I would just like a review from someone on the Coiled distributed team before we hit merge.

idorrington92 commented:
Can we ping someone from Coiled to review this? It's been months now and would be nice to merge this

mrocklin commented:
Hi @idorrington92 sorry that this hasn't gotten much love.  I suspect that most folks are out during the holidays.  I'm holding down the fort for the moment.  

If I were reviewing this initially I probably would have said "the try-except logic around `cluster.wait_for_workers` seems super simple and good to me 👍 .  However, the implementation for `Cluster.wait_for_workers` and the changes to `update_scheduler_info` seem also *probably* good, but like an area that might be error prone.  Do we actually want/need this for some reason?  If not I'd be inclined to skip this and stay with status quo.  It seems to me like the upstream issue was mostly asking for the ability of *downstream* clusters to have their `wait_for_worker` methods respected, not tha we implement one for all `Cluster` classes.  

My preference is to drop the implementation (even if this means dropping the test), and just stick with the try-except logic that you have in `client.py` (which seems super straightforward).  However, I appreciate that you've also gone through several rounds of review, and having expectations change on you after so long probably isn't fun. 

Do you have any thoughts on the above?  Is there a solid reason to implement `wait_for_workers` and `update_scheduler_info` that I'm not seeing here?

mrocklin commented:
Said a different way, I'm more than happy to merge the client.py changes immediately.  I'd want to look over the other changes more thoroughly, which would take me some time.  Or somoene like @jacobtomlinson who has already looked over them could merge if he's feeling confident.

fjetter commented on distributed/deploy/cluster.py:
```diff
@@ -579,3 +588,59 @@ def __eq__(self, other):
+
+    async def update_scheduler_info(self) -> None:
+        """Send comm to scheduler requesting information and update scheduler_info accordingly"""
+        await self._scheduler_info_comm.write({"op": "identity"})
+        self.scheduler_info = SchedulerInfo(await self._scheduler_info_comm.read())
```
You can simply do

`self.scheduler_info = SchedulerInfo(await self._scheduler.identity())`

This will allow this connection to be reused and handles the write/read part of this call. Specifically you will not need to manage the lifecycle of the comm yourself.

FWIW I think the same _should_ be done with the `_watch_worker_status_comm`. I don't see a reason why these calls would deserve a dedicated connection. I assume `_watch_worker_status_comm` was introduced *before* we had a connection pool

idorrington92 replied:
Hi @fjetter, 
I get an attribute error when making this change and using a LocalCluster. 

It was quite a while ago that I wrote this code, but I'm pretty sure the reason I went down the creating-a-comm route was because clusters don't (necessarily?) have a _scheduler attribute. 

Am I missing something? Should there be a _scheduler attribute?

fjetter replied:
Sorry, this should've been `self.scheduler_comm`.

This should always be set if a cluster is already started. The implementation is a bit messy, though. `Cluster` doesn't actually set it but `SpecCluster` does... regardless, having this set is a requirement since we're using this in many other places as well, i.e. it is safe to assume that subclasses implement this as well

idorrington92 replied:
That worked! I've removed all the messy comm stuff now :)
Thank you :)

fjetter commented on distributed/client.py:
```diff
@@ -1383,7 +1383,12 @@ def wait_for_workers(
+        try:
+            return self.cluster.wait_for_workers(n_workers, timeout)
+        except AttributeError:
+            # Most likely, either self.cluster is None, or the cluster has not
+            # implemented a wait_for_workers method
```
nit: You added wait_for_workers to the base class, i.e. every cluster will come equipped with this. I typically prefer dealing with these situations by being explicit, i.e. `if cluster is None; ...`

idorrington92 replied:
I agree, changed it to an if-else

idorrington92 commented:
> Hi @idorrington92 sorry that this hasn't gotten much love. I suspect that most folks are out during the holidays. I'm holding down the fort for the moment.
> 
> If I were reviewing this initially I probably would have said "the try-except logic around `cluster.wait_for_workers` seems super simple and good to me +1 . However, the implementation for `Cluster.wait_for_workers` and the changes to `update_scheduler_info` seem also _probably_ good, but like an area that might be error prone. Do we actually want/need this for some reason? If not I'd be inclined to skip this and stay with status quo. It seems to me like the upstream issue was mostly asking for the ability of _downstream_ clusters to have their `wait_for_worker` methods respected, not tha we implement one for all `Cluster` classes.
> 
> My preference is to drop the implementation (even if this means dropping the test), and just stick with the try-except logic that you have in `client.py` (which seems super straightforward). However, I appreciate that you've also gone through several rounds of review, and having expectations change on you after so long probably isn't fun.
> 
> Do you have any thoughts on the above? Is there a solid reason to implement `wait_for_workers` and `update_scheduler_info` that I'm not seeing here?

Hi @mrocklin,
Thank you for your quick reply :)
I see what you mean. I took the issue quite literally and just implemented the changes that were requested. But perhaps a simple try-catch in client.py would allow clusters to do everything the issue really needs. I've not used clusters much, so don't know from a users perspective whether cluster.wait_for_workers is very useful.

FYI - Following the comment from @fjetter above, I've replaced that try-catch with an if-else, so that will need to be reverted if we do go down that route. 

I'm happy either way, I learned a lot from working on this issue, and would rather remove most of my changes and contribute something useful than add a load of code that'll cause problems later on :) 

I think it's better if people with more user knowledge than me decide which route we go down though :)

jacobtomlinson reviewed (commented):
Sure. Could you fix up the merge conflicts?

idorrington92 commented:
I've handled the merge conflicts. I'm getting some failing tests, but I'm seeing these test failures in other PRs as well so don't think it's to do with my changes...

jacobtomlinson commented:
It looks like `distributed/deploy/tests/test_cluster.py::test_cluster_wait_for_worker` is failing a bunch here so it's certainly related to these changes. Would you mind taking another look?

jacobtomlinson reviewed (commented):
The CI failures seem to be happening consistently with Python 3.8 only.

I had a skim over the code and can't see anything that sticks out as being a syntax problem but maybe you could test locally with 3.8 to find the problem?

idorrington92 commented:
It is only failing in Python 3.8, because of threads leaking. Is this

jacobtomlinson reviewed (approved):
Thanks for fixing this up. CI failures now look unrelated.

Thanks for pushing this along, sorry that it's taken so long to get in.

consideRatio commented:
This was a breaking change for dask-gateway's client i think, because it assumes that its available instead of opting in to it if it is.

Can we complement this PR with a conditional check to verify the new function is available or similar? Possibly emitting a warning or similar?
